File: src/GitHub/ActivityAggregator.cs
namespace GitHubWallpaper.GitHub;

/// <summary>
/// Объединяет события GitHub API и дельты polling в единую ленту активности.
/// </summary>
internal sealed class ActivityAggregator
{
    private const int MaxFeedItems = 30;

    private readonly object _sync = new();
    private readonly Dictionary<string, FeedState> _feeds = new(StringComparer.OrdinalIgnoreCase);

    /// <summary>Обновляет ленту событиями из <c>GET /events</c>.</summary>
    public IReadOnlyList<ActivityFeedItem> IntegrateEvents(
        RepoReference repository,
        IReadOnlyList<RepoEventSnapshot> events)
    {
        var items = events
            .Select(evt => new ActivityFeedItem(
                $"event-{evt.Id}",
                MapEventKind(evt.Type),
                evt.Summary,
                evt.ActorLogin,
                evt.CreatedAt ?? DateTimeOffset.UtcNow,
                evt.HtmlUrl,
                IsNew: false))
            .ToList();

        return MergeFeed(repository, items);
    }

    /// <summary>Добавляет в ленту новые коммиты, обнаруженные при polling.</summary>
    public IReadOnlyList<ActivityFeedItem> IntegrateCommits(
        RepoReference repository,
        IReadOnlyList<RepoCommitSnapshot> commits)
    {
        var state = GetState(repository);
        var deltaItems = new List<ActivityFeedItem>();
        var isBootstrap = state.SeenCommitShas.Count == 0;

        foreach (var commit in commits)
        {
            if (string.IsNullOrWhiteSpace(commit.Sha) || !state.SeenCommitShas.Add(commit.Sha))
            {
                continue;
            }

            if (isBootstrap)
            {
                continue;
            }

            deltaItems.Add(new ActivityFeedItem(
                $"commit-{commit.Sha}",
                "push",
                commit.Message,
                commit.AuthorName,
                commit.AuthorDate ?? DateTimeOffset.UtcNow,
                commit.HtmlUrl,
                IsNew: true));
        }

        if (deltaItems.Count == 0)
        {
            return GetFeed(repository);
        }

        return MergeFeed(repository, deltaItems, prepend: true);
    }

    /// <summary>Добавляет в ленту новые pull request'ы.</summary>
    public IReadOnlyList<ActivityFeedItem> IntegratePulls(
        RepoReference repository,
        IReadOnlyList<RepoPullSnapshot> pulls)
    {
        return IntegrateNumberedItems(
            repository,
            pulls,
            state => state.SeenPullNumbers,
            pull => pull.Number,
            pull => new ActivityFeedItem(
                $"pr-{pull.Number}",
                "pr",
                pull.Title,
                pull.UserLogin,
                pull.CreatedAt ?? DateTimeOffset.UtcNow,
                pull.HtmlUrl,
                IsNew: true));
    }

    /// <summary>Добавляет в ленту новые issues.</summary>
    public IReadOnlyList<ActivityFeedItem> IntegrateIssues(
        RepoReference repository,
        IReadOnlyList<RepoIssueSnapshot> issues)
    {
        return IntegrateNumberedItems(
            repository,
            issues,
            state => state.SeenIssueNumbers,
            issue => issue.Number,
            issue => new ActivityFeedItem(
                $"issue-{issue.Number}",
                "issue",
                issue.Title,
                issue.UserLogin,
                issue.CreatedAt ?? DateTimeOffset.UtcNow,
                issue.HtmlUrl,
                IsNew: true));
    }

    /// <summary>Добавляет в ленту новые релизы.</summary>
    public IReadOnlyList<ActivityFeedItem> IntegrateReleases(
        RepoReference repository,
        IReadOnlyList<RepoReleaseSnapshot> releases)
    {
        return IntegrateNumberedItems(
            repository,
            releases,
            state => state.SeenReleaseIds,
            release => release.Id,
            release => new ActivityFeedItem(
                $"release-{release.Id}",
                "release",
                release.Name,
                release.TagName,
                release.PublishedAt ?? DateTimeOffset.UtcNow,
                release.HtmlUrl,
                IsNew: true));
    }

    /// <summary>Обновляет ленту при смене статуса CI run.</summary>
    public IReadOnlyList<ActivityFeedItem> IntegrateCiRun(
        RepoReference repository,
        RepoCiRunSnapshot? run)
    {
        if (run is null)
        {
            return GetFeed(repository);
        }

        var state = GetState(repository);
        var signature = $"{run.Id}:{run.Status}:{run.Conclusion}";
        if (state.LastCiSignature == signature)
        {
            return GetFeed(repository);
        }

        var isBootstrap = state.LastCiSignature is null;
        state.LastCiSignature = signature;

        if (isBootstrap)
        {
            return GetFeed(repository);
        }

        var title = run.Conclusion switch
        {
            "success" => $"CI passed: {run.Name}",
            "failure" => $"CI failed: {run.Name}",
            "cancelled" => $"CI cancelled: {run.Name}",
            _ => $"CI {run.Status}: {run.Name}",
        };

        var item = new ActivityFeedItem(
            $"ci-{run.Id}-{run.UpdatedAt?.ToUnixTimeSeconds() ?? 0}",
            "ci",
            title,
            run.Conclusion ?? run.Status,
            run.UpdatedAt ?? DateTimeOffset.UtcNow,
            run.HtmlUrl,
            IsNew: true);

        return MergeFeed(repository, [item], prepend: true);
    }

    /// <summary>Текущая лента для репозитория.</summary>
    public IReadOnlyList<ActivityFeedItem> GetFeed(RepoReference repository)
    {
        lock (_sync)
        {
            return _feeds.TryGetValue(repository.Slug, out var state)
                ? state.Items.ToList()
                : [];
        }
    }

    /// <summary>Сбрасывает состояние ленты репозитория.</summary>
    public void Clear(RepoReference repository)
    {
        lock (_sync)
        {
            _feeds.Remove(repository.Slug);
        }
    }

    private IReadOnlyList<ActivityFeedItem> IntegrateNumberedItems<TItem, TKey>(
        RepoReference repository,
        IReadOnlyList<TItem> items,
        Func<FeedState, HashSet<TKey>> seenSelector,
        Func<TItem, TKey> keySelector,
        Func<TItem, ActivityFeedItem> map)
        where TKey : notnull
    {
        var state = GetState(repository);
        var seen = seenSelector(state);
        var deltaItems = new List<ActivityFeedItem>();
        var isBootstrap = seen.Count == 0;

        foreach (var item in items)
        {
            var key = keySelector(item);
            if (!seen.Add(key))
            {
                continue;
            }

            if (isBootstrap)
            {
                continue;
            }

            deltaItems.Add(map(item));
        }

        if (deltaItems.Count == 0)
        {
            return GetFeed(repository);
        }

        return MergeFeed(repository, deltaItems, prepend: true);
    }

    private IReadOnlyList<ActivityFeedItem> MergeFeed(
        RepoReference repository,
        IReadOnlyList<ActivityFeedItem> incoming,
        bool prepend = false)
    {
        lock (_sync)
        {
            var state = GetState(repository);
            var merged = new Dictionary<string, ActivityFeedItem>(StringComparer.Ordinal);

            foreach (var existing in state.Items)
            {
                merged[existing.Id] = existing with { IsNew = false };
            }

            foreach (var item in incoming)
            {
                if (merged.TryGetValue(item.Id, out var existing))
                {
                    merged[item.Id] = item with { IsNew = item.IsNew || existing.IsNew };
                }
                else
                {
                    merged[item.Id] = item;
                }
            }

            var ordered = merged.Values
                .OrderByDescending(item => item.Timestamp)
                .Take(MaxFeedItems)
                .ToList();

            if (prepend)
            {
                var newestIds = incoming.Select(item => item.Id).ToHashSet(StringComparer.Ordinal);
                ordered = ordered
                    .Select(item => item with { IsNew = item.IsNew || newestIds.Contains(item.Id) })
                    .ToList();
            }

            state.Items = ordered;
            return state.Items.ToList();
        }
    }

    private FeedState GetState(RepoReference repository)
    {
        lock (_sync)
        {
            if (!_feeds.TryGetValue(repository.Slug, out var state))
            {
                state = new FeedState();
                _feeds[repository.Slug] = state;
            }

            return state;
        }
    }

    private static string MapEventKind(string type) =>
        type switch
        {
            "PushEvent" => "push",
            "PullRequestEvent" => "pr",
            "IssuesEvent" => "issue",
            "ReleaseEvent" => "release",
            "WatchEvent" => "watch",
            "ForkEvent" => "fork",
            _ => "event",
        };

    private sealed class FeedState
    {
        public List<ActivityFeedItem> Items { get; set; } = [];

        public HashSet<string> SeenCommitShas { get; } = new(StringComparer.Ordinal);

        public HashSet<int> SeenPullNumbers { get; } = [];

        public HashSet<int> SeenIssueNumbers { get; } = [];

        public HashSet<long> SeenReleaseIds { get; } = [];

        public string? LastCiSignature { get; set; }
    }
}

File: python/github_wallpaper/github/activity_aggregator.py
"""Объединение событий API и дельт polling в ленту (ActivityAggregator.cs)."""

from __future__ import annotations

from dataclasses import replace
from datetime import datetime, timezone
from typing import Callable, TypeVar

from github_wallpaper.github.api_parsers import (
    ActivityFeedItem,
    RepoCiRunSnapshot,
    RepoCommitSnapshot,
    RepoEventSnapshot,
    RepoIssueSnapshot,
    RepoPullSnapshot,
    RepoReleaseSnapshot,
)
from github_wallpaper.repo_url_parser import RepoReference

_MAX_FEED_ITEMS = 30

TItem = TypeVar("TItem")
TKey = TypeVar("TKey")


class ActivityAggregator:
    """Единая лента активности по репозиторию."""

    def __init__(self) -> None:
        self._feeds: dict[str, _FeedState] = {}

    def integrate_events(
        self,
        repository: RepoReference,
        events: list[RepoEventSnapshot],
    ) -> list[ActivityFeedItem]:
        items = [
            ActivityFeedItem(
                id=f"event-{event.id}",
                kind=_map_event_kind(event.type),
                title=event.summary,
                subtitle=event.actor_login,
                timestamp=event.created_at or datetime.now(timezone.utc),
                html_url=event.html_url,
                is_new=False,
            )
            for event in events
        ]
        return self._merge_feed(repository, items)

    def integrate_commits(
        self,
        repository: RepoReference,
        commits: list[RepoCommitSnapshot],
    ) -> list[ActivityFeedItem]:
        state = self._get_state(repository)
        delta_items: list[ActivityFeedItem] = []
        is_bootstrap = len(state.seen_commit_shas) == 0

        for commit in commits:
            if not commit.sha or commit.sha in state.seen_commit_shas:
                continue
            state.seen_commit_shas.add(commit.sha)
            if is_bootstrap:
                continue
            delta_items.append(
                ActivityFeedItem(
                    id=f"commit-{commit.sha}",
                    kind="push",
                    title=commit.message,
                    subtitle=commit.author_name,
                    timestamp=commit.author_date or datetime.now(timezone.utc),
                    html_url=commit.html_url,
                    is_new=True,
                )
            )

        if not delta_items:
            return self.get_feed(repository)
        return self._merge_feed(repository, delta_items, prepend=True)

    def integrate_pulls(
        self,
        repository: RepoReference,
        pulls: list[RepoPullSnapshot],
    ) -> list[ActivityFeedItem]:
        return self._integrate_numbered_items(
            repository,
            pulls,
            lambda state: state.seen_pull_numbers,
            lambda pull: pull.number,
            lambda pull: ActivityFeedItem(
                id=f"pr-{pull.number}",
                kind="pr",
                title=pull.title,
                subtitle=pull.user_login,
                timestamp=pull.created_at or datetime.now(timezone.utc),
                html_url=pull.html_url,
                is_new=True,
            ),
        )

    def integrate_issues(
        self,
        repository: RepoReference,
        issues: list[RepoIssueSnapshot],
    ) -> list[ActivityFeedItem]:
        return self._integrate_numbered_items(
            repository,
            issues,
            lambda state: state.seen_issue_numbers,
            lambda issue: issue.number,
            lambda issue: ActivityFeedItem(
                id=f"issue-{issue.number}",
                kind="issue",
                title=issue.title,
                subtitle=issue.user_login,
                timestamp=issue.created_at or datetime.now(timezone.utc),
                html_url=issue.html_url,
                is_new=True,
            ),
        )

    def integrate_releases(
        self,
        repository: RepoReference,
        releases: list[RepoReleaseSnapshot],
    ) -> list[ActivityFeedItem]:
        return self._integrate_numbered_items(
            repository,
            releases,
            lambda state: state.seen_release_ids,
            lambda release: release.id,
            lambda release: ActivityFeedItem(
                id=f"release-{release.id}",
                kind="release",
                title=release.name,
                subtitle=release.tag_name,
                timestamp=release.published_at or datetime.now(timezone.utc),
                html_url=release.html_url,
                is_new=True,
            ),
        )

    def integrate_ci_run(
        self,
        repository: RepoReference,
        run: RepoCiRunSnapshot | None,
    ) -> list[ActivityFeedItem]:
        if run is None:
            return self.get_feed(repository)

        state = self._get_state(repository)
        signature = f"{run.id}:{run.status}:{run.conclusion}"
        if state.last_ci_signature == signature:
            return self.get_feed(repository)

        is_bootstrap = state.last_ci_signature is None
        state.last_ci_signature = signature
        if is_bootstrap:
            return self.get_feed(repository)

        if run.conclusion == "success":
            title = f"CI passed: {run.name}"
        elif run.conclusion == "failure":
            title = f"CI failed: {run.name}"
        elif run.conclusion == "cancelled":
            title = f"CI cancelled: {run.name}"
        else:
            title = f"CI {run.status}: {run.name}"

        updated_ts = int(run.updated_at.timestamp()) if run.updated_at else 0
        item = ActivityFeedItem(
            id=f"ci-{run.id}-{updated_ts}",
            kind="ci",
            title=title,
            subtitle=run.conclusion or run.status,
            timestamp=run.updated_at or datetime.now(timezone.utc),
            html_url=run.html_url,
            is_new=True,
        )
        return self._merge_feed(repository, [item], prepend=True)

    def get_feed(self, repository: RepoReference) -> list[ActivityFeedItem]:
        state = self._feeds.get(repository.slug.lower())
        if state is None:
            return []
        return list(state.items)

    def clear(self, repository: RepoReference) -> None:
        self._feeds.pop(repository.slug.lower(), None)

    def _integrate_numbered_items(
        self,
        repository: RepoReference,
        items: list[TItem],
        seen_selector: Callable[[_FeedState], set[TKey]],
        key_selector: Callable[[TItem], TKey],
        map_item: Callable[[TItem], ActivityFeedItem],
    ) -> list[ActivityFeedItem]:
        state = self._get_state(repository)
        seen = seen_selector(state)
        delta_items: list[ActivityFeedItem] = []
        is_bootstrap = len(seen) == 0

        for item in items:
            key = key_selector(item)
            if key in seen:
                continue
            seen.add(key)
            if is_bootstrap:
                continue
            delta_items.append(map_item(item))

        if not delta_items:
            return self.get_feed(repository)
        return self._merge_feed(repository, delta_items, prepend=True)

    def _merge_feed(
        self,
        repository: RepoReference,
        incoming: list[ActivityFeedItem],
        *,
        prepend: bool = False,
    ) -> list[ActivityFeedItem]:
        state = self._get_state(repository)
        merged: dict[str, ActivityFeedItem] = {}

        for existing in state.items:
            merged[existing.id] = replace(existing, is_new=False)

        for item in incoming:
            if item.id in merged:
                previous = merged[item.id]
                merged[item.id] = replace(item, is_new=item.is_new or previous.is_new)
            else:
                merged[item.id] = item

        ordered = sorted(merged.values(), key=lambda entry: entry.timestamp, reverse=True)[:_MAX_FEED_ITEMS]

        if prepend:
            newest_ids = {item.id for item in incoming}
            ordered = [
                replace(item, is_new=item.is_new or item.id in newest_ids) for item in ordered
            ]

        state.items = ordered
        return list(state.items)

    def _get_state(self, repository: RepoReference) -> _FeedState:
        key = repository.slug.lower()
        state = self._feeds.get(key)
        if state is None:
            state = _FeedState()
            self._feeds[key] = state
        return state


class _FeedState:
    def __init__(self) -> None:
        self.items: list[ActivityFeedItem] = []
        self.seen_commit_shas: set[str] = set()
        self.seen_pull_numbers: set[int] = set()
        self.seen_issue_numbers: set[int] = set()
        self.seen_release_ids: set[int] = set()
        self.last_ci_signature: str | None = None


def _map_event_kind(event_type: str) -> str:
    mapping = {
        "PushEvent": "push",
        "PullRequestEvent": "pr",
        "IssuesEvent": "issue",
        "ReleaseEvent": "release",
        "WatchEvent": "watch",
        "ForkEvent": "fork",
    }
    return mapping.get(event_type, "event")
